fmlcore/src/main/java/net/minecraftforge/fml/VersionChecker.java

```diff
@@ -116,46 +119,47 @@ public void run()
                     return;
                 }
 
+                client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(HTTP_TIMEOUT_SECS)).build();
                 gatherMods().forEach(this::process);
             }
 
             /**
-             * Opens stream for given URL while following redirects
+             * Returns the response body as a String for the given URL while following redirects
              */
-            private InputStream openUrlStream(URL url) throws IOException
-            {
+            private String openUrlString(URL url) throws IOException, URISyntaxException, InterruptedException {
                 URL currentUrl = url;
                 for (int redirects = 0; redirects < MAX_HTTP_REDIRECTS; redirects++)
                 {
-                    URLConnection c = currentUrl.openConnection();
-                    if (c instanceof HttpURLConnection)
+                    var request = HttpRequest.newBuilder()
+                            .uri(currentUrl.toURI())
+                            .timeout(Duration.ofSeconds(HTTP_TIMEOUT_SECS))
+                            .GET()
+                            .build();
+
+                    HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+
+                    int responseCode = response.statusCode();
+                    if (responseCode >= 300 && responseCode <= 399)
                     {
-                        HttpURLConnection huc = (HttpURLConnection) c;
-                        huc.setInstanceFollowRedirects(false);
-                        int responseCode = huc.getResponseCode();
-                        if (responseCode >= 300 && responseCode <= 399)
+                        try
                         {
-                            try
-                            {
-                                String loc = huc.getHeaderField("Location");
-                                currentUrl = new URL(currentUrl, loc);
-                                continue;
-                            }
-                            finally
-                            {
-                                huc.disconnect();
-                            }
+                            String newLocation = response.headers().map().get("Location").get(0);
+                            currentUrl = new URL(currentUrl, newLocation);
+                            continue;
+                        }
+                        catch (NullPointerException e)
+                        {
+                            throw new IOException("Got a 3xx response code but Location header was null while trying to fetch " + url);
                         }
                     }
 
-                    return c.getInputStream();
+                    return response.body();
```

Comment:
Shouldn't we shutdown the client executor thread and connections if it is no longer referenced?

Reply:
HttpClient doesn't provide a way to do so.

Reply:
If I remember correctly, connections are automatically closed either once a response has been received, it times out or when the server has closed it.

The HttpClient object is reused for each mod's update check to improve performance and reduce memory usage.